### PR TITLE
Revert "change arg back to a TypeSet since it supports multiple values and order doesn't matter"

### LINKED ISSUE #21203: aws_eks_node_group.instance_types not respecting order for instance_types
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Affected Resource(s)

* aws_eks_node_group

### Terraform Configuration Files

Please include all Terraform configurations required to reproduce the bug. Bug reports without a functional reproduction may be closed without investigation.

```hcl
resource "aws_eks_node_group" "eks_cluster_node_group_autoscaler" {
  ...
  instance_types  = [
    "m6i.2xlarge",
    "m5.2xlarge",
    "m4.2xlarge",
    "t3.2xlarge"
  ]
  ...
}
```

### Expected Behavior

Instance types should be set in the order given to the node_group. This ensures order of instances for the autoscaling group and lets AWS fallback in the correct order.

### Actual Behavior

List of instances is converted to a Set internally (I guess to ensure uniqueness?) and the given order is lost.

```
# module.eks.aws_eks_node_group.test will be created
  + resource "aws_eks_node_group" "test" {
      + ami_type               = (known after apply)
      + arn                    = (known after apply)
      + capacity_type          = (known after apply)
      + cluster_name           = (known after apply)
      + disk_size              = (known after apply)
      + id                     = (known after apply)
      + instance_types         = [
          + "m4.2xlarge",
          + "m5.2xlarge",
          + "m6i.2xlarge",
          + "t3.2xlarge",
        ]
      + node_group_name        = (known after apply)
      + node_group_name_prefix = (known after apply)
      + node_role_arn          = "<ARN>"
      + release_version        = (known after apply)
      + resources              = (known after apply)
      + status                 = (known after apply)
      + subnet_ids             = [
          <SUBNETS>
        ]
      + version                = "1.21"

      + launch_template {
          + id      = (known after apply)
          + name    = (known after apply)
          + version = (known after apply)
        }

      + scaling_config {
          + desired_size = 10
          + max_size     = 20
          + min_size     = 1
        }

      + update_config {
          + max_unavailable            = (known after apply)
          + max_unavailable_percentage = (known after apply)
        }
    }
```

### Steps to Reproduce

1. `terraform apply`

### PR DESCRIPTION
**Pull request by sylr**
This reverts commit eff27751df66602298bc1e73352f188f8d1f3982.

Fixes #21203.
Relates #20532.

Original PR stated that the order of instance types did not matter but it absolutely does.

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccXXX'

...
```

### CODE REVIEW

**Comment by sylr** [comment]
@ewbankkit could you please review this ?

**Comment by ewbankkit** [comment]
The [CloudFormation Resource Schema](https://github.com/aws-cloudformation/cloudformation-resource-schema) definition for `AWS::EKS::Nodegroup` contains

```
    "InstanceTypes": {
      "type": "array",
      "uniqueItems": false,
      "items": {
        "type": "string"
      }
    },
 ```
 
 and as the default value of `insertionOrder` [is `true`](https://github.com/aws-cloudformation/cloudformation-resource-schema#insertionorder), the `instance_types` attribute should be `TypeList`.

**Comment by basert** [comment]
I think thats exactly what this MR tries to solve. It was changed to `TypeSet` in a previous release, this is the revert back to `TypeList` again.

**Comment by ewbankkit** [comment]
Will wait for @anGie44 to take a look before merging.

File: internal/service/eks/node_group.go [hunk 1]
```diff
@@ -313,8 +313,8 @@ func resourceNodeGroupCreate(ctx context.Context, d *schema.ResourceData, meta i
 		input.DiskSize = aws.Int64(int64(v.(int)))
 	}
 
-	if v := d.Get("instance_types").(*schema.Set); v.Len() > 0 {
-		input.InstanceTypes = flex.ExpandStringSet(v)
+	if v := d.Get("instance_types").([]interface{}); len(v) > 0 {
+		input.InstanceTypes = flex.ExpandStringList(v)
```
Comment (anGie44) [inline internal/service/eks/node_group.go:317]: To account for empty values: 
```suggestion
	if v, ok := d.GetOk("instance_types"); ok && len(v.([]interface{})) > 0 && v.([]interface{})[0] != nil {
		input.InstanceTypes = flex.ExpandStringList(v.([]interface{}))
```

**Review by anGie44 — APPROVED** [review]
Thanks for this fix @sylr 🚀 ! Preemptively approving while the acceptance tests finish running on our side.

**Comment by anGie44** [comment]
Just a heads up: i've come across a test failure pertaining to the order of `instance_types` configured vs. returned from the API (will investigate shortly)
```
------- Stdout: -------
=== RUN   TestAccEKSNodeGroup_InstanceTypes_multiple
=== PAUSE TestAccEKSNodeGroup_InstanceTypes_multiple
=== CONT  TestAccEKSNodeGroup_InstanceTypes_multiple
    node_group_test.go:285: Step 1/2 error: After applying this test step, the plan was not empty.
        stdout:
        
        
        An execution plan has been generated and is shown below.
        Resource actions are indicated with the following symbols:
        -/+ destroy and then create replacement
        
        Terraform will perform the following actions:
        
          # aws_eks_node_group.test must be replaced
        -/+ resource "aws_eks_node_group" "test" {
              ~ ami_type               = "AL2_x86_64" -> (known after apply)
              ~ arn                    = "arn:aws:eks:us-west-2:*******:nodegroup/tf-acc-test-7522170719420147764/tf-acc-test-7522170719420147764/64be8539-e106-834b-23e2-55faaba1e52e" -> (known after apply)
              ~ capacity_type          = "ON_DEMAND" -> (known after apply)
              ~ disk_size              = 20 -> (known after apply)
              ~ id                     = "tf-acc-test-7522170719420147764:tf-acc-test-7522170719420147764" -> (known after apply)
              ~ instance_types         = [
                  - "t2.medium",
                    "t3.large",
                    "t3.medium",
                  + "t2.medium",
                    "t2.large",
                ]
              + node_group_name_prefix = (known after apply)
              ~ release_version        = "1.21.4-20211013" -> (known after apply)
              ~ resources              = [
                  - {
                      - autoscaling_groups              = [
                          - {
                              - name = "eks-tf-acc-test-7522170719420147764-64be8539-e106-834b-23e2-55faaba1e52e"
                            },
                        ]
                      - remote_access_security_group_id = ""
                    },
                ] -> (known after apply)
              ~ status                 = "ACTIVE" -> (known after apply)
              ~ tags_all               = {} -> (known after apply)
              ~ version                = "1.21" -> (known after apply)
                # (4 unchanged attributes hidden)
        
        
              ~ update_config {
                  ~ max_unavailable            = 1 -> (known after apply)
                  ~ max_unavailable_percentage = 0 -> (known after apply)
                }
                # (1 unchanged block hidden)
            }
        
        Plan: 1 to add, 0 to change, 1 to destroy.
    testing_new.go:70: Error running post-test destroy, there may be dangling resources: exit status 1
        
        Error: Error deleting VPC: DependencyViolation: The vpc 'vpc-0f0d760e484a322a4' has dependencies and cannot be deleted.
        	status code: 400, request id: 58925730-e376-40ea-86c7-01347bf2e1bc
        
        
--- FAIL: TestAccEKSNodeGroup_InstanceTypes_multiple (1534.42s)
```

**Comment by anGie44** [comment]
Confirming test is now passing with change:
```
--- PASS: TestAccEKSNodeGroup_InstanceTypes_multiple (1292.25s)
```

**Pull request merged**